WebGL "is resource" checking and object validation contains repeated code

kkinnunen-apple opened the pull request:
#### f13b4beff03458f7ee54332a23507cd35e22ad35
<pre>
WebGL &quot;is resource&quot; checking and object validation contains repeated code
<a href="https://bugs.webkit.org/show_bug.cgi?id=260475">https://bugs.webkit.org/show_bug.cgi?id=260475</a>
rdar://114207432

Reviewed by Dan Glastonbury.

Checks of &quot;isTexture()&quot; and others, and webgl object validation has
following problems:
 - Uses inheritance for customizing behavior, forcing the base class
   have state which is not common to all subclasses
 - Has custom but similar code for each resource, with varying level
   of consistency (slight checking differences)
 - Has duplicated isContextLost() checks inside validation functions

Fix these by:
 - Using templates to implement the common logic for validation
   instead of using WebGLObject base class. This allows future changes
   to remove the virtualization, further simplifying the code.
 - Expose the differences in resource behavior in the `isUsable()`,
   `isInitialized()` methods in the resources.
 - Implement WebGLBindingPoint&lt;T&gt; template to implement &quot;this object
   was bound&quot; marker in a consistent, &quot;provably&quot; not missing any cases
   kind of way.
 - Removes unused WebGLRenderbuffer::isInitialized() and related
   WebGLFramebuffer::WebGLAttachment::isInitialized() code.
   The code was unused and appropriate to remove here due to the name
   clash with the added isInitialized().
 - Change validateWebGLObject object argument from pointer to object
   since it was always called with an existing object.

* Source/WebCore/html/canvas/EXTDisjointTimerQuery.cpp:
(WebCore::EXTDisjointTimerQuery::isQueryEXT):
(WebCore::EXTDisjointTimerQuery::beginQueryEXT):
(WebCore::EXTDisjointTimerQuery::queryCounterEXT):
(WebCore::EXTDisjointTimerQuery::getQueryObjectEXT):
* Source/WebCore/html/canvas/EXTDisjointTimerQueryWebGL2.cpp:
(WebCore::EXTDisjointTimerQueryWebGL2::queryCounterEXT):
* Source/WebCore/html/canvas/OESVertexArrayObject.cpp:
(WebCore::OESVertexArrayObject::isVertexArrayOES):
(WebCore::OESVertexArrayObject::bindVertexArrayOES):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::validateAndCacheBufferBinding):
(WebCore::WebGL2RenderingContext::bindFramebuffer):
(WebCore::WebGL2RenderingContext::framebufferTextureLayer):
(WebCore::WebGL2RenderingContext::getFragDataLocation):
(WebCore::WebGL2RenderingContext::deleteQuery):
(WebCore::WebGL2RenderingContext::isQuery):
(WebCore::WebGL2RenderingContext::beginQuery):
(WebCore::WebGL2RenderingContext::getQueryParameter):
(WebCore::WebGL2RenderingContext::isSampler):
(WebCore::WebGL2RenderingContext::bindSampler):
(WebCore::WebGL2RenderingContext::samplerParameteri):
(WebCore::WebGL2RenderingContext::samplerParameterf):
(WebCore::WebGL2RenderingContext::getSamplerParameter):
(WebCore::WebGL2RenderingContext::isSync):
(WebCore::WebGL2RenderingContext::clientWaitSync):
(WebCore::WebGL2RenderingContext::waitSync):
(WebCore::WebGL2RenderingContext::getSyncParameter):
(WebCore::WebGL2RenderingContext::isTransformFeedback):
(WebCore::WebGL2RenderingContext::bindTransformFeedback):
(WebCore::WebGL2RenderingContext::transformFeedbackVaryings):
(WebCore::WebGL2RenderingContext::getTransformFeedbackVarying):
(WebCore::WebGL2RenderingContext::setIndexedBufferBinding):
(WebCore::WebGL2RenderingContext::getUniformIndices):
(WebCore::WebGL2RenderingContext::getActiveUniforms):
(WebCore::WebGL2RenderingContext::getUniformBlockIndex):
(WebCore::WebGL2RenderingContext::getActiveUniformBlockParameter):
(WebCore::WebGL2RenderingContext::getActiveUniformBlockName):
(WebCore::WebGL2RenderingContext::uniformBlockBinding):
(WebCore::WebGL2RenderingContext::isVertexArray):
(WebCore::WebGL2RenderingContext::bindVertexArray):
(WebCore::WebGL2RenderingContext::getFramebufferAttachmentParameter):
(WebCore::WebGL2RenderingContext::getParameter):
* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGLBuffer.h:
(WebCore::WebGLBuffer::didBind):
* Source/WebCore/html/canvas/WebGLDebugShaders.cpp:
(WebCore::WebGLDebugShaders::getTranslatedShaderSource):
* Source/WebCore/html/canvas/WebGLFramebuffer.cpp:
(WebCore::WebGLFramebuffer::WebGLFramebuffer):
* Source/WebCore/html/canvas/WebGLFramebuffer.h:
* Source/WebCore/html/canvas/WebGLObject.h:
(WebCore::WebGLBindingPoint::WebGLBindingPoint):
(WebCore::WebGLBindingPoint::operator=):
(WebCore::WebGLBindingPoint::operator== const):
(WebCore::WebGLBindingPoint::operator bool const):
(WebCore::WebGLBindingPoint::get const):
(WebCore::WebGLBindingPoint::operator-&gt; const):
(WebCore::WebGLBindingPoint::operator* const):
(WebCore::WebGLBindingPoint::operator RefPtr&lt;T&gt; const):
(WebCore::WebGLBindingPoint::didBind):
(WebCore::WebGLObject::isDeleted const):
(WebCore::objectOrZero):
(WebCore::WebGLObject::isDeleted): Deleted.
* Source/WebCore/html/canvas/WebGLProgram.h:
* Source/WebCore/html/canvas/WebGLQuery.h:
* Source/WebCore/html/canvas/WebGLRenderbuffer.cpp:
(WebCore::WebGLRenderbuffer::WebGLRenderbuffer):
* Source/WebCore/html/canvas/WebGLRenderbuffer.h:
* Source/WebCore/html/canvas/WebGLRenderingContext.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::clearIfComposited):
(WebCore::WebGLRenderingContextBase::reshape):
(WebCore::WebGLRenderingContextBase::attachShader):
(WebCore::WebGLRenderingContextBase::bindAttribLocation):
(WebCore::WebGLRenderingContextBase::validateAndCacheBufferBinding):
(WebCore::WebGLRenderingContextBase::bindBuffer):
(WebCore::WebGLRenderingContextBase::bindFramebuffer):
(WebCore::WebGLRenderingContextBase::bindRenderbuffer):
(WebCore::WebGLRenderingContextBase::bindTexture):
(WebCore::WebGLRenderingContextBase::compileShader):
(WebCore::WebGLRenderingContextBase::detachShader):
(WebCore::WebGLRenderingContextBase::getActiveAttrib):
(WebCore::WebGLRenderingContextBase::getActiveUniform):
(WebCore::WebGLRenderingContextBase::getAttachedShaders):
(WebCore::WebGLRenderingContextBase::getAttribLocation):
(WebCore::WebGLRenderingContextBase::getParameter):
(WebCore::WebGLRenderingContextBase::getProgramParameter):
(WebCore::WebGLRenderingContextBase::getProgramInfoLog):
(WebCore::WebGLRenderingContextBase::getShaderParameter):
(WebCore::WebGLRenderingContextBase::getShaderInfoLog):
(WebCore::WebGLRenderingContextBase::getShaderSource):
(WebCore::WebGLRenderingContextBase::getUniform):
(WebCore::WebGLRenderingContextBase::getUniformLocation):
(WebCore::WebGLRenderingContextBase::isBuffer):
(WebCore::WebGLRenderingContextBase::isFramebuffer):
(WebCore::WebGLRenderingContextBase::isProgram):
(WebCore::WebGLRenderingContextBase::isRenderbuffer):
(WebCore::WebGLRenderingContextBase::isShader):
(WebCore::WebGLRenderingContextBase::isTexture):
(WebCore::WebGLRenderingContextBase::linkProgram):
(WebCore::WebGLRenderingContextBase::linkProgramWithoutInvalidatingAttribLocations):
(WebCore::WebGLRenderingContextBase::shaderSource):
(WebCore::WebGLRenderingContextBase::useProgram):
(WebCore::WebGLRenderingContextBase::validateProgram):
(WebCore::WebGLRenderingContextBase::setFramebuffer):
(WebCore::WebGLRenderingContextBase::validateNullableWebGLObject): Deleted.
(WebCore::WebGLRenderingContextBase::validateWebGLObject): Deleted.
(WebCore::WebGLRenderingContextBase::validateWebGLProgramOrShader): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
(WebCore::WebGLRenderingContextBase::validateWebGLObject):
(WebCore::WebGLRenderingContextBase::validateNullableWebGLObject):
(WebCore::WebGLRenderingContextBase::validateIsWebGLObject const):
* Source/WebCore/html/canvas/WebGLSampler.h:
* Source/WebCore/html/canvas/WebGLShader.h:
* Source/WebCore/html/canvas/WebGLSync.h:
* Source/WebCore/html/canvas/WebGLTexture.cpp:
(WebCore::WebGLTexture::WebGLTexture):
(WebCore::WebGLTexture::didBind):
(WebCore::WebGLTexture::setTarget): Deleted.
* Source/WebCore/html/canvas/WebGLTexture.h:
* Source/WebCore/html/canvas/WebGLTimerQueryEXT.h:
* Source/WebCore/html/canvas/WebGLTransformFeedback.h:
* Source/WebCore/html/canvas/WebGLVertexArrayObjectBase.h:
(WebCore::WebGLVertexArrayObjectBase::didBind):
(WebCore::WebGLVertexArrayObjectBase::isUsable const):
(WebCore::WebGLVertexArrayObjectBase::isInitialized const):
(WebCore::WebGLVertexArrayObjectBase::hasEverBeenBound const): Deleted.
(WebCore::WebGLVertexArrayObjectBase::setHasEverBeenBound): Deleted.
* Source/WebCore/inspector/InspectorShaderProgram.cpp:
(WebCore::InspectorShaderProgram::updateShader):

Canonical link: <a href="https://commits.webkit.org/267121@main">https://commits.webkit.org/267121@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ec66d2a8e729f2f91e2a573013e67d5f80e929f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15985 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16358 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17435 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14712 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15859 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18958 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16087 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/17243 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15863 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16340 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13352 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18181 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13601 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14154 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14601 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14320 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17576 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14909 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12645 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14163 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3757 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18527 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14728 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/html/canvas/WebGLRenderbuffer.h:
```diff
@@ -40,7 +40,6 @@ class WebGLRenderbuffer final : public WebGLObject {
     void setInternalFormat(GCGLenum internalformat)
     {
         m_internalFormat = internalformat;
-        m_initialized = false;
```
Is this a "poor man's `std::optional`"? (ie. was `m_initialized` referring to whether `m_internalFormat` has been set? Should `m_internalFormat` be a `std::optional<GCGLenum>`?

kkinnunen-apple replied:
No, the code was unused and being removed here, not added. I'm unsure what it was originally for.

djg commented on Source/WebCore/html/canvas/WebGLObject.h:
```diff
@@ -90,7 +135,8 @@ class WebGLObject : public RefCounted<WebGLObject> {
 };
 
-inline PlatformGLObject objectOrZero(WebGLObject* object)
+template<typename T>
+PlatformGLObject objectOrZero(const T& object)
```
Now we've gone from one function to a specialised version for every derived type of WebGLObject. If this is inlined, that's probably not such a big deal, but it is necessary?

kkinnunen-apple replied:
Well, we do invoke RefPtr templates, specialized for 80% of types in WebKit, for 80% of the lines that WebKit has..

djg commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
```diff
@@ -1067,6 +1062,41 @@ class WebGLRenderingContextBase : public GraphicsContextGL::Client, public GPUBa
     WeakPtrFactory<WebGLRenderingContextBase> m_contextObjectWeakPtrFactory;
 };
 
+template<typename T>
+bool WebGLRenderingContextBase::validateWebGLObject(const char* functionName, const T& object)
```
Should SFINAE be used to only enable these functions for types `T` that are derived from `WebGLObject`?

kkinnunen-apple replied:
Well, the idea is to be able to remove the dependency to the base class..
I think the other idea is that if you *want* to validate with non-WebGLObject, that working could be seen as a feature.
That being said, passing wrong object tends to give bad error messages so from that perspective some typing would be good..

djg commented on Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
```diff
@@ -1721,8 +1719,10 @@ std::optional<WebGL2RenderingContext::ActiveQueryKey> WebGL2RenderingContext::va
 
 void WebGL2RenderingContext::beginQuery(GCGLenum target, WebGLQuery& query)
 {
+    if (isContextLost())
+        return;
```
Instead of writing `if (isContextLost()) return;` in all functions, should this check be performed at the JS binding level and the binding generator write the code for us? Then, we'd never forgot to write the check.

kkinnunen-apple replied:
Yeah, that would be a good idea perhaps. There's couple of cases that need to be handled manually -- these are in fact probably marked by WebGL spec IDL as WebGLHandlesContextLoss or similar annotation.

djg commented on Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
```diff
@@ -3111,7 +3133,7 @@ WebGLAny WebGL2RenderingContext::getParameter(GCGLenum pname)
         return m_boundTransformFeedbackBuffer;
     case GraphicsContextGL::TRANSFORM_FEEDBACK_BINDING:
-        return m_boundTransformFeedback == m_defaultTransformFeedback ? nullptr : m_boundTransformFeedback;
+        return m_boundTransformFeedback == m_defaultTransformFeedback ? nullptr : RefPtr { m_boundTransformFeedback.get() };
```
Does the `operator RefPtr<T>` on `WebGLBindingPoint<T>` not kick in here?

kkinnunen-apple replied:
I don't know how it would be possible since a) the return value is WebGLAny and b) tertiary operator returning nullptr
e.g. the typing is
`WebGLAny = nullptr_t or WebGLBindingPoint<WebGLTransformFeedback, ...> `
I think there's too many conversions to reach the RefPtr< WebGLTransformFeedback> inside WebGLAny, but I never remember all the intricacies of the implicit conversions. Certainly it had a compile error..

djg reviewed (approved):
:shipit:

kkinnunen-apple committed:
WebGL "is resource" checking and object validation contains repeated code
https://bugs.webkit.org/show_bug.cgi?id=260475
rdar://114207432

Reviewed by Dan Glastonbury.

Checks of "isTexture()" and others, and webgl object validation has
following problems:
 - Uses inheritance for customizing behavior, forcing the base class
   have state which is not common to all subclasses
 - Has custom but similar code for each resource, with varying level
   of consistency (slight checking differences)
 - Has duplicated isContextLost() checks inside validation functions

Fix these by:
 - Using templates to implement the common logic for validation
   instead of using WebGLObject base class. This allows future changes
   to remove the virtualization, further simplifying the code.
 - Expose the differences in resource behavior in the `isUsable()`,
   `isInitialized()` methods in the resources.
 - Implement WebGLBindingPoint<T> template to implement "this object
   was bound" marker in a consistent, "provably" not missing any cases
   kind of way.
 - Removes unused WebGLRenderbuffer::isInitialized() and related
   WebGLFramebuffer::WebGLAttachment::isInitialized() code.
   The code was unused and appropriate to remove here due to the name
   clash with the added isInitialized().
 - Change validateWebGLObject object argument from pointer to object
   since it was always called with an existing object.

* Source/WebCore/html/canvas/EXTDisjointTimerQuery.cpp:
(WebCore::EXTDisjointTimerQuery::isQueryEXT):
(WebCore::EXTDisjointTimerQuery::beginQueryEXT):
(WebCore::EXTDisjointTimerQuery::queryCounterEXT):
(WebCore::EXTDisjointTimerQuery::getQueryObjectEXT):
* Source/WebCore/html/canvas/EXTDisjointTimerQueryWebGL2.cpp:
(WebCore::EXTDisjointTimerQueryWebGL2::queryCounterEXT):
* Source/WebCore/html/canvas/OESVertexArrayObject.cpp:
(WebCore::OESVertexArrayObject::isVertexArrayOES):
(WebCore::OESVertexArrayObject::bindVertexArrayOES):
* Source/WebCore/html/canvas/WebGL2RenderingContext.cpp:
(WebCore::WebGL2RenderingContext::validateAndCacheBufferBinding):
(WebCore::WebGL2RenderingContext::bindFramebuffer):
(WebCore::WebGL2RenderingContext::framebufferTextureLayer):
(WebCore::WebGL2RenderingContext::getFragDataLocation):
(WebCore::WebGL2RenderingContext::deleteQuery):
(WebCore::WebGL2RenderingContext::isQuery):
(WebCore::WebGL2RenderingContext::beginQuery):
(WebCore::WebGL2RenderingContext::getQueryParameter):
(WebCore::WebGL2RenderingContext::isSampler):
(WebCore::WebGL2RenderingContext::bindSampler):
(WebCore::WebGL2RenderingContext::samplerParameteri):
(WebCore::WebGL2RenderingContext::samplerParameterf):
(WebCore::WebGL2RenderingContext::getSamplerParameter):
(WebCore::WebGL2RenderingContext::isSync):
(WebCore::WebGL2RenderingContext::clientWaitSync):
(WebCore::WebGL2RenderingContext::waitSync):
(WebCore::WebGL2RenderingContext::getSyncParameter):
(WebCore::WebGL2RenderingContext::isTransformFeedback):
(WebCore::WebGL2RenderingContext::bindTransformFeedback):
(WebCore::WebGL2RenderingContext::transformFeedbackVaryings):
(WebCore::WebGL2RenderingContext::getTransformFeedbackVarying):
(WebCore::WebGL2RenderingContext::setIndexedBufferBinding):
(WebCore::WebGL2RenderingContext::getUniformIndices):
(WebCore::WebGL2RenderingContext::getActiveUniforms):
(WebCore::WebGL2RenderingContext::getUniformBlockIndex):
(WebCore::WebGL2RenderingContext::getActiveUniformBlockParameter):
(WebCore::WebGL2RenderingContext::getActiveUniformBlockName):
(WebCore::WebGL2RenderingContext::uniformBlockBinding):
(WebCore::WebGL2RenderingContext::isVertexArray):
(WebCore::WebGL2RenderingContext::bindVertexArray):
(WebCore::WebGL2RenderingContext::getFramebufferAttachmentParameter):
(WebCore::WebGL2RenderingContext::getParameter):
* Source/WebCore/html/canvas/WebGL2RenderingContext.h:
* Source/WebCore/html/canvas/WebGLBuffer.h:
(WebCore::WebGLBuffer::didBind):
* Source/WebCore/html/canvas/WebGLDebugShaders.cpp:
(WebCore::WebGLDebugShaders::getTranslatedShaderSource):
* Source/WebCore/html/canvas/WebGLFramebuffer.cpp:
(WebCore::WebGLFramebuffer::WebGLFramebuffer):
* Source/WebCore/html/canvas/WebGLFramebuffer.h:
* Source/WebCore/html/canvas/WebGLObject.h:
(WebCore::WebGLBindingPoint::WebGLBindingPoint):
(WebCore::WebGLBindingPoint::operator=):
(WebCore::WebGLBindingPoint::operator== const):
(WebCore::WebGLBindingPoint::operator bool const):
(WebCore::WebGLBindingPoint::get const):
(WebCore::WebGLBindingPoint::operator-> const):
(WebCore::WebGLBindingPoint::operator* const):
(WebCore::WebGLBindingPoint::operator RefPtr<T> const):
(WebCore::WebGLBindingPoint::didBind):
(WebCore::WebGLObject::isDeleted const):
(WebCore::objectOrZero):
(WebCore::WebGLObject::isDeleted): Deleted.
* Source/WebCore/html/canvas/WebGLProgram.h:
* Source/WebCore/html/canvas/WebGLQuery.h:
* Source/WebCore/html/canvas/WebGLRenderbuffer.cpp:
(WebCore::WebGLRenderbuffer::WebGLRenderbuffer):
* Source/WebCore/html/canvas/WebGLRenderbuffer.h:
* Source/WebCore/html/canvas/WebGLRenderingContext.h:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::clearIfComposited):
(WebCore::WebGLRenderingContextBase::reshape):
(WebCore::WebGLRenderingContextBase::attachShader):
(WebCore::WebGLRenderingContextBase::bindAttribLocation):
(WebCore::WebGLRenderingContextBase::validateAndCacheBufferBinding):
(WebCore::WebGLRenderingContextBase::bindBuffer):
(WebCore::WebGLRenderingContextBase::bindFramebuffer):
(WebCore::WebGLRenderingContextBase::bindRenderbuffer):
(WebCore::WebGLRenderingContextBase::bindTexture):
(WebCore::WebGLRenderingContextBase::compileShader):
(WebCore::WebGLRenderingContextBase::detachShader):
(WebCore::WebGLRenderingContextBase::getActiveAttrib):
(WebCore::WebGLRenderingContextBase::getActiveUniform):
(WebCore::WebGLRenderingContextBase::getAttachedShaders):
(WebCore::WebGLRenderingContextBase::getAttribLocation):
(WebCore::WebGLRenderingContextBase::getParameter):
(WebCore::WebGLRenderingContextBase::getProgramParameter):
(WebCore::WebGLRenderingContextBase::getProgramInfoLog):
(WebCore::WebGLRenderingContextBase::getShaderParameter):
(WebCore::WebGLRenderingContextBase::getShaderInfoLog):
(WebCore::WebGLRenderingContextBase::getShaderSource):
(WebCore::WebGLRenderingContextBase::getUniform):
(WebCore::WebGLRenderingContextBase::getUniformLocation):
(WebCore::WebGLRenderingContextBase::isBuffer):
(WebCore::WebGLRenderingContextBase::isFramebuffer):
(WebCore::WebGLRenderingContextBase::isProgram):
(WebCore::WebGLRenderingContextBase::isRenderbuffer):
(WebCore::WebGLRenderingContextBase::isShader):
(WebCore::WebGLRenderingContextBase::isTexture):
(WebCore::WebGLRenderingContextBase::linkProgram):
(WebCore::WebGLRenderingContextBase::linkProgramWithoutInvalidatingAttribLocations):
(WebCore::WebGLRenderingContextBase::shaderSource):
(WebCore::WebGLRenderingContextBase::useProgram):
(WebCore::WebGLRenderingContextBase::validateProgram):
(WebCore::WebGLRenderingContextBase::setFramebuffer):
(WebCore::WebGLRenderingContextBase::validateNullableWebGLObject): Deleted.
(WebCore::WebGLRenderingContextBase::validateWebGLObject): Deleted.
(WebCore::WebGLRenderingContextBase::validateWebGLProgramOrShader): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
(WebCore::WebGLRenderingContextBase::validateWebGLObject):
(WebCore::WebGLRenderingContextBase::validateNullableWebGLObject):
(WebCore::WebGLRenderingContextBase::validateIsWebGLObject const):
* Source/WebCore/html/canvas/WebGLSampler.h:
* Source/WebCore/html/canvas/WebGLShader.h:
* Source/WebCore/html/canvas/WebGLSync.h:
* Source/WebCore/html/canvas/WebGLTexture.cpp:
(WebCore::WebGLTexture::WebGLTexture):
(WebCore::WebGLTexture::didBind):
(WebCore::WebGLTexture::setTarget): Deleted.
* Source/WebCore/html/canvas/WebGLTexture.h:
* Source/WebCore/html/canvas/WebGLTimerQueryEXT.h:
* Source/WebCore/html/canvas/WebGLTransformFeedback.h:
* Source/WebCore/html/canvas/WebGLVertexArrayObjectBase.h:
(WebCore::WebGLVertexArrayObjectBase::didBind):
(WebCore::WebGLVertexArrayObjectBase::isUsable const):
(WebCore::WebGLVertexArrayObjectBase::isInitialized const):
(WebCore::WebGLVertexArrayObjectBase::hasEverBeenBound const): Deleted.
(WebCore::WebGLVertexArrayObjectBase::setHasEverBeenBound): Deleted.
* Source/WebCore/inspector/InspectorShaderProgram.cpp:
(WebCore::InspectorShaderProgram::updateShader):

Canonical link: https://commits.webkit.org/267121@main
